scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+            unur_set_stream(self._messages.handle)
+
+            for i in range(size):
+                out[i] = unur_sample_cont(rng)
+                if PyErr_Occurred():
+                    return
+
+            msg = self._messages.get()
+            if msg:
+                raise UNURANError(msg)
+        finally:
+            _lock.release()
+            if self._callback_wrapper is not None:
+                release_unuran_callback(&callback)
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    cdef inline void _rvs_discr(self, int[::1] out) except *:
+        """
+        Sample random variates from a discrete distribution.
+
+        Parameters
+        ----------
+        out : int[::1]
+            A memory view of size ``size`` to store the result.
+        """
+        cdef:
+            ccallback_t callback
+            unur_gen *rng = self.rng
+            size_t i
+            size_t size = len(out)
+
+        if self._callback_wrapper is not None:
+            init_unuran_callback(&callback, self._callback_wrapper,
+                                 self.params)
+        _lock.acquire()
+        try:
+            self._messages.clear()
+            unur_set_stream(self._messages.handle)
+
+            for i in range(size):
+                out[i] = unur_sample_discr(rng)
+                if PyErr_Occurred():
+                    return
+
+            msg = self._messages.get()
+            if msg:
+                raise UNURANError(msg)
+        finally:
+            _lock.release()
+            if self._callback_wrapper is not None:
+                release_unuran_callback(&callback)
+
+    def rvs(self, size=None):
+        """
+        Sample from the distribution.
+
+        Parameters
+        ----------
+        size : int or tuple, optional
+            The shape of samples. Default is ``None`` in which case a scalar
+            sample is returned.
+
+        Returns
+        -------
+        rvs : array_like
+            A NumPy array of random variates.
+        """
+        cdef double[::1] out_cont
+        cdef int[::1] out_discr
+        N = 1 if size is None else np.prod(size)
+        if unur_distr_is_cont(unur_get_distr(self.rng)):
+            out_cont = np.empty(N, dtype=np.float64)
+            self._rvs_cont(out_cont)
+            if size is None:
+                return out_cont[0]
+            return np.asarray(out_cont).reshape(size)
+        elif unur_distr_is_discr(unur_get_distr(self.rng)):
+            out_discr = np.empty(N, dtype=np.int32)
+            self._rvs_discr(out_discr)
+            if size is None:
+                return out_discr[0]
+            return np.asarray(out_discr).reshape(size)
+        else:
+            raise NotImplementedError("only univariate continuous and "
+                                      "discrete distributions supported")
+
+    @property
+    def seed(self):
+        return self.numpy_rng
+
+    @seed.setter
+    def seed(self, new_seed):
```

Comment:
This should not be named `seed`. At this point, `self.numpy_rng` is no longer a seed. `numpy_rng` is probably the correct name. The setter should probably be a proper method that someone calls rather than a `@property` setter. You could probably just declare the attribute `cdef readonly object numpy_rng` and change this one into a (documented) method. `set_numpy_rng()` would be a reasonable name.

Reply:
Actually, you probably don't want to expose this attribute for reading at all.

Reply:
I agree that a documented method to set the numpy_rng would be better. Though, about naming: wouldn't that introduce too many naming conventions within SciPy itself: `seed` for QMC, `random_state` for SciPy's distributions, and `numpy_rng` for UNU.RAN generators?

Reply:
`seed` for QMC was probably a mistake (mine, I think), as I've suggested elsewhere. I would say `rng` is the general direction we're going in for the name of this kind of input (as we move away from `RandomState`), though obviously that's overused in this context. `numpy_rng` is a reasonable qualifier for this context.

Reply:
> `numpy_rng` is a reasonable qualifier for this context.

+1 for `numpy_rng` (as opposed to `random_state` and `seed`). What do you think @chrisb83, @mckib2, and @tupui?

Reply:
I would prefer just `rng` but fine with me.

Reply:
I am fine with `rng` and `numpy_rng`. If we call it `numpy_rng`, it assumes that numpy will provide the uniform random numbers. In principle, we could also support other URNG.

Reply:
Changed the name from `seed` to `numpy_rng`. Updated docs, tuts, tests accordingly. Also, I removed the getter for the seed and changed the setter to a method: `set_numpy_rng`

Reply:
Just one more vote for plain old `rng`, which would not be ambiguous if it means we can have some consistency in the name of this parameter. (I hope we don't intend to have `numpy_rng` here and `rng` elsewhere? Really, my vote would be to just live with `random_state`, unless we plan on standardizing something else across the board in, for example, SciPy 2.0)

Reply:
I'm fine with `rng` if folks don't think it's confusing here.

Reply:
I am OK with both `rng` and `numpy_rng`. According to @rkern: 

> I would say `rng` is the general direction we're going in for the name of this kind of input (as we move away from `RandomState`), though obviously that's overused in this context. `numpy_rng` is a reasonable qualifier for this context.

IMO, if we go for `numpy_rng` now, then we should do that everywhere else in SciPy.

> Really, my vote would be to just live with `random_state`, unless we plan on standardizing something else across the board in, for example, SciPy 2.0

I think `random_state` alludes to `RandomState` in NumPy which is legacy now. Considering that, we should definitely move to something else like `rng`, `numpy_rng`, or `seed`. In long term, standardizing one of these names would be very beneficial. Though, I guess it's not a topic for this PR. We might want to open a new issue and discuss it on the mailing list.

Reply:
> I think random_state alludes to RandomState in NumPy which is legacy now

Yes, if you know what a `RandomState` is. My impression is that people have been comfortable with `np.random.seed` for a long time, and with that, you might never run across `RandomState` explicitly, even if you've been relying on it. Even if they do know of `RandomState`, `random_state` is also a reasonable English name for the parameter, even if it now accepts `Generator`s.

We have two options, each with pros and cons. Personally, I see having a more consistent interface (with `random_state` across the board) as being better overall, even with the risk to users who are 1) informed enough to know what `RandomState` is but 2) not informed enough to know that `Generator` is preferred.

There have been several (somewhat sweeping) efforts to break all ties with `RandomState`. This seems to be another instance of that. I am worried that with the enthusiasm for that, other things (like a stable SciPy interface) are getting put aside. I'm don't want to speak too strongly - so again, feel free to resolve - but I thought it was worth mentioning.

Reply:
Those are valid points. We could also go with `random_state` for consistency until a decision is reached how to proceed for `rv_continuous` etc It seems difficult to reach a consensus and we should avoid that Tirth has to change the name more than one last time :D

Should we have a final vote `random_state` vs `rng`? I am fine with either option and would go with the majority.

Reply:
The semantics here are different from the existing uses of `random_state`, which coerce the inputs to a `RandomState` and makes use of the global PRNG. Here, we coerce to a `Generator` except in the short-term case where we are using numpy 1.16 where it's unavailable. Importantly, if the passed value is `None`, we get a fresh `Generator` unassociated with any other rather than reintroducing a global PRNG. That's a subtle but important difference in semantics, which is why I want to signal that with a new name rather than reuse the old convention with different semantics alongside the older instances of the convention with the old semantics.

Reply:
@rkern That's a valid point - the behavior is different, so perhaps it deserves a different name.

But we need to weigh this in with the disadvantages of complicating the SciPy interface. Does the subtle but important difference in behavior warrant a name change? 

My guess - and it is not a very educated guess - is that most users won't care whether `RandomState` or `Generator` is used to generate the random numbers. They just need the numbers to be pseudorandom and they need the results to be repeatable. Better scientists will take extra steps, like using large, arbitrary seeds, and they would be pleased to know that under the hood the superior `Generator` is being used, even if the name is still `random_state`. Some people will be confused because they are deeply invested in which algorithm is being used to generate the random numbers, and they would be surprised to find that the numbers are generated differently than they would have expected, given that the name is `random_state`, which has traditionally been associated with `RandomState`. I expect that is a minority, though, and my guess is that these are the people who would read the fine print (the documentation of `random_state` for this class).

It's possible I'm just projecting my opinions onto users, but I figure it's somewhat warranted because I'm more deeply invested in the workings of the scientific Python ecosystem than the average user. If this is my view of random number generation, I don't imagine that the masses will be more detail-oriented than me.

Reply:
I'm less concerned about the differences between `RandomState` and `Generator` per se so much as the differences between `random_state=None` and `rng=None`.

Reply:
I was going to let others chime in, but it's been a while, so I'll go for it.

I'm not sure how changing the name of the parameter is going to alert people to the change in default behavior. It's the default, after all.

Is the thought that this will protect users who pass `random_state=None` because it will raise an error, and then they'd notice that the parameter name is `rng` and that the behavior has changed?

Reply:
> 'd would go with @chrisb83's suggestion [here](https://github.com/scipy/scipy/pull/14215#discussion_r687174721)

It seems like the most sensible thing to do for now. It's best if the name change is done in a separate PR all over SciPy.

Reply:
To continue the discussion for other refactoring, I propose to use the existing #14322. It would be good to have this for an eventual 2.0. Or we can start a more general issue with list of things for 2.0.

Reply:
I think @mdhaber, @chrisb83, and @tupui agree to stay with `random_state` for now. Should I rename it to `random_state` then, @rkern?

Reply:
I'm firmly against `random_state`. Naming the argument differently is a signpost for those who at least skim the docstring to know that it's different. There is nothing I can do for people who don't read documentation at all. But reusing the `random_state` name will mean that people will skim the docstring and think they understand it as like the other uses of `random_state` without reading the one little clause in the middle of that lengthy argument description (that otherwise looks identical) to see what the difference is. Changing the name puts that difference up front.

Reply:
If you do want to use `random_state` and postpone things until a coordinated transition a la #14322, then you will need to use `scipy._lib._util.check_random_state()` to process the argument (you can then extract out the `BitGenerator` and wrap it with a `Generator` for actual use if you want; the important thing is the `check_random_state(None)` behavior).

Reply:
> I'm firmly against `random_state`. Naming the argument differently is a signpost for those who at least skim the docstring to know that it's different.

I agree with your point. But I think that we should stay with `random_state` in this PR and change all instances of `random_state` to `rng`/`numpy_rng` in some other PR and run a deprecation cycle if necessary. If we rush the decision now, we might end up in another unpleasant situation later on. If you disagree, I think it makes sense to keep `numpy_rng`?

Reply:
> If you do want to use `random_state` and postpone things until a coordinated transition a la #14322, then you will need to use `scipy._lib._util.check_random_state()` to process the argument (you can then extract out the `BitGenerator` and wrap it with a `Generator` for actual use if you want; the important thing is the `check_random_state(None)` behavior).

OK, let's do that!

Reply:
If I understand correctly, this means using the old `random_state` behavior in the meantime, and that' sounds good to me.

Reply:
Done: changed the name to `random_state` and used `check_random_state` for any given seed to replicate the behavior of `random_state` elsewhere. All tutorials, tests, and docs updated accordingly.